File: src/main/java/com/paragon/client/ui/configuration/panel/panel/Panel.java
package com.paragon.client.ui.configuration.panel.panel;

import com.paragon.client.ui.configuration.panel.element.Element;

public abstract class Panel extends Element {

    private float lastX;
    private float lastY;

    public Panel(float x, float y, float width, float height) {
        super(0, x, y, width, height);
    }

    public float getLastX() {
        return lastX;
    }

    public void setLastX(float lastX) {
        this.lastX = lastX;
    }

    public float getLastY() {
        return lastY;
    }

    public void setLastY(float lastY) {
        this.lastY = lastY;
    }

}

File: src/main/kotlin/com/paragon/impl/ui/configuration/shared/Panel.kt
package com.paragon.impl.ui.configuration.shared

import com.paragon.impl.ui.util.Click
import net.minecraft.util.math.Vec2f

/**
 * @author Surge
 */
abstract class Panel(x: Float, y: Float, width: Float, height: Float) : RawElement(x, y, width, height) {

    lateinit var title: String
    private var dragging = false

    private var lastPosition = Vec2f(0f, 0f)

    override fun draw(mouseX: Float, mouseY: Float, mouseDelta: Int) {
        if (dragging) {
            x = mouseX - lastPosition.x
            y = mouseY - lastPosition.y
        }
    }

    override fun mouseClicked(mouseX: Float, mouseY: Float, click: Click) {
        super.mouseClicked(mouseX, mouseY, click)

        if (click == Click.LEFT) {
            if (mouseX in x..x + width && mouseY in y..y + height) {
                dragging = true
                lastPosition = Vec2f(mouseX - x, mouseY - y)
            }
        }
    }

    override fun mouseReleased(mouseX: Float, mouseY: Float, click: Click) {
        super.mouseReleased(mouseX, mouseY, click)

        if (click == Click.LEFT) {
            dragging = false
        }
    }

}